test/distributed/fsdp/test_fsdp_meta.py

```diff
@@ -0,0 +1,315 @@
+    def __init__(self, device):
+        super().__init__()
+        self.lin1 = MyLinear(2, 2, bias=False, device=device)
+        self.lin2 = MyLinear(2, 2, bias=False, device=device)
+
+    def forward(self, x):
+        return self.lin2(self.lin1(x))
+
+    def reset_parameters(self, *args, **kwargs):
+        for m in [self.lin1, self.lin2]:
+            if not isinstance(m, FSDP):
+                m.reset_parameters()
+
+
+class NestedModel(nn.Module):
+    def __init__(self, device):
+        super().__init__()
+        torch.manual_seed(0)
+        torch.cuda.manual_seed(0)
+        self.lin1 = MyLinear(2, 2, bias=False, device=device)
+        self.lin1 = wrap(self.lin1)
+        self.lin2 = MyLinear(2, 2, bias=False, device=device)
+        self.l3 = MyModel(device=device)
+        self.l3 = wrap(self.l3)
+
+    def forward(self, x):
+        return self.l3(self.lin2(self.lin1(x)))
+
+    def reset_parameters(self):
+        for m in [self.lin1, self.lin2, self.l3]:
+            if not isinstance(m, FSDP):
+                m.reset_parameters()
+
+def _init_with_reset_params(module):
+    """
+    to_empty + reset_parameters() init function example for modules
+    initailized with device="meta"
+    """
+    is_meta = any(t.is_meta for t in module.parameters())
+    if is_meta:
+        module.to_empty(device=torch.cuda.current_device())
+    with torch.no_grad():
+        module.reset_parameters()
+
+def _init_with_torchdistx(module):
+    """
+    torchdistx-based deferred module initialization function example
+    using ``materialize_module``.
+    """
+    assert _TORCHDISTX_AVAIL
+
+    def check_fn(k):
+        return not isinstance(k, FSDP)
+
+    deferred_init.materialize_module(module, check_fn=check_fn)
+
+class TestFSDPWithMetaDevice(FSDPTest):
+    @property
+    def world_size(self):
+        return 2
+
+    @property
+    def process_group(self):
+        return dist.distributed_c10d._get_default_group()
+
+    def _compare_fsdp(self, fsdp1, fsdp2):
+        with fsdp1.summon_full_params(fsdp1):
+            with fsdp2.summon_full_params(fsdp2):
+                for p1, p2 in zip(fsdp1.parameters(), fsdp2.parameters()):
+                    self.assertTrue(torch.allclose(p1, p2), f"{p1} vs {p2}")
+
+    def _test_simple_model_with_meta_device(self, meta_module_fn, init_fn=None):
+        # Create model on meta device and wrap with FSDP.
+        model = meta_module_fn()
+        is_meta = next(model.parameters()).is_meta
+        fsdp_meta = FSDP(
+            model,
+            auto_wrap_policy=always_wrap,
+            param_init_fn=init_fn,
+        )
+
+        meta_opt = torch.optim.SGD(fsdp_meta.parameters(), lr=1e-3)
+
+        # Test to make sure it is the same model parameters as regular FSDP
+        # approach.
+        regular = MyModel(device="cuda")
+        # For torchdistx init, we don't need to call reset_params, as
+        # deferred_init(model).materialize() is equivalent to model().
```

Comment:
nit:
"`equivalent to model()`"
What is `model()` referring to? Should it be `model.reset_parameters()`?

Reply:
Oh, I meant that 

deferred_init(model).materialize() is equivalent to just calling model() i.e. creating the model with no additional `reset_params`.